Introduce new predicates

alexander-demicev opened the pull request:
<!--
Label the PR with the kind of change this for:

kind/feature
kind/bug
kind/documentation
kind/regression
kind/*
-->

**What this PR does / why we need it**:

Add new predicates to filter out clusters that don't have a control plane ready or were not marked for import. I didn't remove all similar logic from reconciler because I've seen in the past predicates failing and unexpected behavior happening because of that. 

<!-- Enter a description of the change and why this change is needed -->

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [x] squashed commits into logical changes
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

Danil-Grigorev commented on util/predicates/cluster_predicates.go:
```diff
@@ -43,3 +48,82 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+
+	cluster, ok := obj.(*clusterv1.Cluster)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Cluster but got a %T", obj)) // it should never happen
```
Don’t panic here. Return false, this just means that some other object came from cache, which could potentially happen.

richardcase replied:
I agree, maybe log an error and return false?

Danil-Grigorev commented on internal/controllers/import_controller.go:
```diff
@@ -273,38 +273,6 @@ func (r *CAPIImportReconciler) reconcileNormal(ctx context.Context, capiCluster
-	// Check CAPI cluster for label first
-	hasLabel, autoImport := shouldImport(capiCluster)
-	if hasLabel && autoImport {
-		log.V(2).Info("Cluster contains import annotation")
```
It is no longer an annotation, but a label

furkatgofurov7 replied:
It is a removed code, so should not matter?

Danil-Grigorev commented on internal/controllers/import_controller.go:
```diff
@@ -66,10 +64,12 @@ func (r *CAPIImportReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Ma
+		predicates.ResourceHasFilterLabel(log, r.WatchFilterValue),
+		turtlespredicates.ClusterWithoutImportedAnnotation(log),
+		turtlespredicates.ClusterWithReadyControlPlane(log),
+		turtlespredicates.ClusterOrNamespaceWithImportLabel(ctx, log, r.Client, importLabelName),
```
This will no longer work. Namespace is a different object, and predicates.All is an equivalent of && here. Each of the previous predicates are checking if the object is a Cluster object.

richardcase replied:
This also checks if its a Cluster object .

furkatgofurov7 replied:
+1

alexander-demicev replied:
it works as expected, the predicate checks if object is a cluster and gets namespace if label is not found

richardcase commented on util/predicates/cluster_predicates.go:
```diff
@@ -43,3 +48,82 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+
+	cluster, ok := obj.(*clusterv1.Cluster)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Cluster but got a %T", obj)) // it should never happen
```
Same comment as @Danil-Grigorev above with removing the use of the panic

richardcase commented on util/predicates/cluster_predicates_test.go:
```diff
@@ -1,79 +1,148 @@
+		Expect(result).To(BeTrue())
+	})
+
+	It("should return true if client fails to get namespace", func() {
```
Are you thinking of transient errors here?

alexander-demicev replied:
yes

richardcase commented on util/util.go:
```diff
@@ -0,0 +1,60 @@
+package util
```
Could we change the package name to something different? `util` is very generic.

furkatgofurov7 replied:
I think it was me who introduced a `util` when moving predicates to its own package in https://github.com/rancher-sandbox/rancher-turtles/pull/50/commits/6c01783876743422e93be6e8e4cf675a6f4ee708 following on how CAPI does it: https://github.com/kubernetes-sigs/cluster-api/tree/main/util

furkatgofurov7 commented on util/predicates/cluster_predicates_test.go:
```diff
@@ -1,79 +1,148 @@
-					Annotations: map[string]string{},
-				},
+			capiCluster.Annotations = map[string]string{
+				"some-other-annoations": "true",
```
```suggestion
				"some-random-annotation": "true",
```

furkatgofurov7 commented on util/predicates/cluster_predicates_test.go:
```diff
@@ -1,79 +1,148 @@
+		Expect(result).To(BeTrue())
+	})
+
+	It("should return false when cluster has no ready control plane", func() {
```
```suggestion
	It("should return false when cluster does not have ready control plane", func() {
```

alexander-demicev commented:
@furkatgofurov7 @Danil-Grigorev @richardcase all fixed, please take a look again

alexander-demicev commented:
@Danil-Grigorev let's merge e2e suite first :D

furkatgofurov7 commented:
> @Danil-Grigorev let's merge e2e suite first :D

@alexander-demicev can we update the PR, e2e should be merged now

richardcase commented:
Apart from the action failure this looks good to me 👍

Danil-Grigorev commented on internal/controllers/import_controller.go:
```diff
@@ -83,10 +83,12 @@ func (r *CAPIImportReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Ma
 	}
 
-	// TODO: we want the control plane initialized but removing for the time being
-	// capiPredicates := predicates.All(log, predicates.ClusterControlPlaneInitialized(log), predicates.ResourceHasFilterLabel(log, r.WatchFilterValue))
```
This comment had a better handling of predicates. Specifically it used ResourceHasFilterLabel, which used label selector. This functionality is imported from controller-runtime and does not require to be tested, so this decreases the amount of code to review. Additionally this gives confidence in what is being added with the change. Why it is not being used here?

Danil-Grigorev replied:
This is the predicate covering 50% of what is being done within the custom logic we have - https://github.com/kubernetes-sigs/controller-runtime/blob/1449c365465511e047333f9bc504e47cb0fae1ae/pkg/predicate/predicate.go#L352

alexander-demicev replied:
That's not going to work in this case. We have to check the cluster label anyway in the predicate because we use the namespace as a fallback option: if the cluster doesn't have a label, look at the namespace. If we remove the label check for the cluster in the predicate, then when the namespace is not labeled, it's always going to return false.

Danil-Grigorev replied:
Yeah, maybe not in the current form. This requires a separate set of predicates on each resource to function properly, and ideally a mechanism to pass these predicates to event mapper.

alexander-demicev commented:
I managed to run e2e locally and it passed

furkatgofurov7 commented on util/predicates/cluster_predicates.go:
```diff
@@ -59,3 +63,84 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+	}
+
+	if !cluster.Status.ControlPlaneReady {
+		log.V(4).Info("Cluster does not have an ready control plane, will not attempt to map resource")
```
```suggestion
		log.V(4).Info("Cluster does not have a ready control plane, will not attempt to map resource")
```

furkatgofurov7 commented on util/predicates/cluster_predicates.go:
```diff
@@ -59,3 +63,84 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+		return false
+	}
+
+	log.V(6).Info("Cluster has an ready control plane, will attempt to map resource")
```
```suggestion
	log.V(6).Info("Cluster has a ready control plane, will attempt to map resource")
```

furkatgofurov7 commented on util/predicates/cluster_predicates.go:
```diff
@@ -59,3 +63,84 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+	}
+}
+
+func processIfClusterReadyControlPlane(logger logr.Logger, obj client.Object) bool {
```
can we add a godoc as we have for other funcs?

furkatgofurov7 commented on util/predicates/cluster_predicates.go:
```diff
@@ -59,3 +63,84 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+
+	shouldImport, err := util.ShouldAutoImport(ctx, log, cl, cluster, label)
+	if err != nil {
+		log.Error(err, "Client failed to get namespace, this error is ignored returning true")
```
this is not always the case, `ShouldAutoImport` can return true also when there is an import annotation set on the cluster object. Reading this log does not help much tbh, I would suggest rather having something like:

```suggestion
		log.Error(err, "namespace or cluster has already import annotation set, ignoring it")
```

furkatgofurov7 commented on util/predicates/cluster_predicates_test.go:
```diff
@@ -17,79 +17,148 @@ limitations under the License.
 	})
 	Context("when CAPI cluster has no annotation", func() {
 		It("should return true", func() {
-			capiCluster := &clusterv1.Cluster{
-				ObjectMeta: metav1.ObjectMeta{
-					Annotations: map[string]string{
-						"some-other-annoations": "true",
-					},
-				},
-			}
-			capiCluster = &clusterv1.Cluster{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "test-cluster",
-					Namespace: "test-ns",
-				},
-			}
```
yeah this is not needed, but I would rather leave it for readiness of the code

furkatgofurov7 commented on util/predicates/cluster_predicates_test.go:
```diff
@@ -17,79 +17,148 @@ limitations under the License.
 	})
 	Context("when CAPI cluster has a random annotation", func() {
 		It("should return true", func() {
-			capiCluster := &clusterv1.Cluster{
-				ObjectMeta: metav1.ObjectMeta{
-					Annotations: map[string]string{"some-random-annoation": "true"},
-				},
-			}
-			capiCluster = &clusterv1.Cluster{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:        "test-cluster",
-					Namespace:   "test-ns",
-					Annotations: map[string]string{},
-				},
```
same as above

furkatgofurov7 reviewed (commented):
Few nits, overall lgtm:

alexander-demicev commented:
@furkatgofurov7 Thanks for the review, I addressed all comments except for the last ones about tests. I like using before each blocks to set up objects for tests and then only assign parameters that are different inside it blocks. I'm not strongly against repeating object setup but it just looks prettier to me with before each.
